pkg/controller/webhook/webhook_controller.go

```diff
@@ -0,0 +1,170 @@
+		glog.Errorf(`HTTP server Shutdown: %v`, err)
+	}
+}
+
+func (c *Controller) initializeHandlers() *http.ServeMux {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/validate", adaptHandler(c.validateHandlerFunc))
+	return mux
+}
+
+// adaptHandler wraps an admission review function to be consumed through HTTP.
+func adaptHandler(handler func(*admission_v1beta1.AdmissionReview) *admission_v1beta1.AdmissionResponse) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var body []byte
+		if r.Body != nil {
+			if data, err := ioutil.ReadAll(r.Body); err == nil {
+				body = data
+			}
+		}
+
+		if len(body) == 0 {
+			http.Error(w, "empty body", http.StatusBadRequest)
+			return
+		}
+
+		contentType := r.Header.Get("Content-Type")
+
+		if contentType != "application/json" {
+			http.Error(w, "invalid Content-Type, expect `application/json`", http.StatusUnsupportedMediaType)
+			return
+		}
+
+		var admissionResponse *admission_v1beta1.AdmissionResponse
+		ar := admission_v1beta1.AdmissionReview{}
+		if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+			admissionResponse = &admission_v1beta1.AdmissionResponse{
+				Result: &meta_v1.Status{
+					Message: err.Error(),
+				},
+			}
+		} else {
+			admissionResponse = handler(&ar)
+		}
+
+		admissionReview := admission_v1beta1.AdmissionReview{}
+		if admissionResponse != nil {
+			admissionReview.Response = admissionResponse
+			if ar.Request != nil {
+				admissionReview.Response.UID = ar.Request.UID
+			}
+		}
+
+		resp, err := json.Marshal(admissionReview)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+		}
+
+		if _, err := w.Write(resp); err != nil {
+			http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+		}
+	}
+}
+
+func (c *Controller) validateHandlerFunc(review *admission_v1beta1.AdmissionReview) *admission_v1beta1.AdmissionResponse {
+	request := review.Request
+	var err error
+
+	switch request.Kind.Kind {
+	case "Application":
+		var application shipper.Application
+		err = json.Unmarshal(request.Object.Raw, &application)
+	case "Release":
+		var release shipper.Release
+		err = json.Unmarshal(request.Object.Raw, &release)
+	case "Cluster":
+		var cluster shipper.Cluster
+		err = json.Unmarshal(request.Object.Raw, &cluster)
+	case "InstallationTarget":
+		var installationTarget shipper.InstallationTarget
+		err = json.Unmarshal(request.Object.Raw, &installationTarget)
+	case "CapacityTarget":
+		var capacityTarget shipper.CapacityTarget
+		err = json.Unmarshal(request.Object.Raw, &capacityTarget)
+	case "TrafficTarget":
+		var trafficTarget shipper.TrafficTarget
+		err = json.Unmarshal(request.Object.Raw, &trafficTarget)
+	}
+
+	if err != nil {
+		return &admission_v1beta1.AdmissionResponse{
+			Result: &meta_v1.Status{
+				Message: err.Error(),
+			},
+		}
+	}
+
+	return &admission_v1beta1.AdmissionResponse{
```

Comment:
I don't understand this. The only thing that happens here is we make sure objects are syntactically correct. But the apiserver does that already.
Are we missing a chunk of functionality here? Like, this is where we should be checking for rollout blocks and such.

Reply:
Right now, our internal version of Kubernetes doesn't check this. So if someone were to create an Application object with the wrong indentation, what happens now is that the Application ends up in the queue, our code fails to decode it, and it ends up blocking the queue until it's fixed.

You're right though; this is also where we will put the logic to take action later, like roll out blocks.